Expose `NonEmptyTuple`

Similar to #659, `NonEmptyTuple` is a convenient type to expose:

https://github.com/sindresorhus/type-fest/blob/f63c3437bc483e7532698d14548494802c04f372/source/internal.d.ts#L90-L93

We could augment it to accept a generic type argument:

```ts
type NonEmptyTuple<TupleType = unknown> = readonly [TupleType, ...TupleType[]];
```

<!-- POLAR PLEDGE BADGE START -->
## Upvote & Fund

- We're using [Polar.sh](https://polar.sh/sindresorhus) so you can upvote and help fund this issue.
- The funding will be given to active contributors.
- Thank you in advance for helping prioritize & fund our backlog.

<a href="https://polar.sh/sindresorhus/type-fest/issues/661">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/sindresorhus/type-fest/issues/661/pledge.svg?darkmode=1">
  <img alt="Fund with Polar" src="https://polar.sh/api/github/sindresorhus/type-fest/issues/661/pledge.svg">
</picture>
</a>
<!-- POLAR PLEDGE BADGE END -->

Hm, maybe it's not such a good idea? You can't use `array.length` checks to guard against empty arrays, and sparse arrays add a lot of edge cases: https://github.com/microsoft/TypeScript/issues/38000 ([SO thread](https://stackoverflow.com/questions/56006111/is-it-possible-to-define-a-non-empty-array-type-in-typescript))

```ts
function printFirstElement<Tuple extends NonEmptyTuple>(tuple: Tuple) {
	console.log(tuple[0]);
}

printFirstElement(['a', 'b', 'c']);
//=> 'a'

printFirstElement([]);
//=> error TS2345: Argument of type '[]' is not assignable to parameter of type 'NonEmptyTuple'.

declare const arr: string[];

if (arr.length > 0 {
	printFirstElement(arr);
	//=> error TS2345: Argument of type 'string[]' is not assignable to parameter of type 'NonEmptyTuple'.
}
```

--- a/index.d.ts
+++ b/index.d.ts
@@ -129,6 +129,7 @@ export type {IsNull} from './source/is-null';
 export type {IfNull} from './source/if-null';
 export type {And} from './source/and';
 export type {Or} from './source/or';
+export type {NonEmptyTuple} from './source/non-empty-tuple';
 
 // Template literal types
 export type {CamelCase} from './source/camel-case';

--- a/readme.md
+++ b/readme.md
@@ -201,6 +201,7 @@ Click the type names for complete docs.
 - [`DistributedPick`](source/distributed-pick.d.ts) - Picks keys from a type, distributing the operation over a union.
 - [`And`](source/and.d.ts) - Returns a boolean for whether two given types are both true.
 - [`Or`](source/or.d.ts) - Returns a boolean for whether either of two given types are true.
+- [`NonEmptyTuple`](source/non-empty-tuple.d.ts) - Matches any non-empty tuple.
 
 ### Type Guard
 

--- a/source/internal.d.ts
+++ b/source/internal.d.ts
@@ -252,11 +252,6 @@ Matches any unknown array or tuple.
 export type UnknownArrayOrTuple = readonly [...unknown[]];
 
 /**
-Matches any non empty tuple.
-*/
-export type NonEmptyTuple = readonly [unknown, ...unknown[]];
-
-/**
 Returns a boolean for whether the two given types extends the base type.
 */
 export type IsBothExtends<BaseType, FirstType, SecondType> = FirstType extends BaseType

--- a/source/merge-deep.d.ts
+++ b/source/merge-deep.d.ts
@@ -6,9 +6,9 @@ import type {
 	ArrayTail,
 	FirstArrayElement,
 	IsBothExtends,
-	NonEmptyTuple,
 	UnknownArrayOrTuple,
 } from './internal';
+import type {NonEmptyTuple} from './non-empty-tuple';
 import type {UnknownRecord} from './unknown-record';
 import type {EnforceOptional} from './enforce-optional';
 import type {SimplifyDeep} from './simplify-deep';

--- a/source/non-empty-tuple.d.ts
+++ b/source/non-empty-tuple.d.ts
@@ -1,0 +1,21 @@
+/**
+Matches any non-empty tuple.
+
+@example
+```
+import type {NonEmptyTuple} from 'type-fest';
+
+const sum = (...numbers: NonEmptyTuple<number>) => numbers.reduce((total, value) => total + value, 0);
+
+sum(1, 2, 3);
+//=> 6
+
+sum();
+//=> Error: Expected at least 1 arguments, but got 0.
+```
+
+@see {@link RequireAtLeastOne} for objects
+
+@category Array
+*/
+export type NonEmptyTuple<T = unknown> = readonly [T, ...T[]];

--- a/test-d/non-empty-tuple.ts
+++ b/test-d/non-empty-tuple.ts
@@ -1,0 +1,10 @@
+import {expectType} from 'tsd';
+import type {NonEmptyTuple} from '../index';
+
+declare const sum: (...numbers: NonEmptyTuple<number>) => number;
+
+expectType<number>(sum(1, 2, 3));
+expectType<number>(sum(1));
+
+// @ts-expect-error
+sum();

